plutus-tx/src/PlutusTx/Data/List.hs

```diff
@@ -98,13 +100,64 @@ instance (UnsafeFromData a, Pretty a) => Pretty (List a) where
     {-# INLINEABLE pretty #-}
     pretty = pretty . toSOP
 
-toBuiltinList :: List a -> BuiltinList BuiltinData
-toBuiltinList = coerce
-{-# INLINEABLE toBuiltinList #-}
+instance Semigroup (List a) where
+    (<>) = append
+    {-# INLINEABLE (<>) #-}
 
-fromBuiltinList :: BuiltinList BuiltinData -> List a
-fromBuiltinList = coerce
-{-# INLINEABLE fromBuiltinList #-}
+instance Monoid (List a) where
+    mempty = nil
+    {-# INLINEABLE mempty #-}
+
+instance Haskell.Semigroup  (List a) where
+    (<>) = append
+    {-# INLINEABLE (<>) #-}
+
+instance Haskell.Monoid (List a) where
+    mempty = nil
+    {-# INLINEABLE mempty #-}
+
+{- Note [Making the cons case non-strict in caseList and caseList']
+
+A function parameter's strictness is in theory irrelevant to the Plinth compiler.
+However, it does affect how GHC compiles the function's callsites to GHC Core.
+
+In this instance of `caseList`/`caseList'`, if `c` is strict, is passed a function
+that calls `caseList`/`caseList'`, GHC will end up creating two mutually recursive
+bindings for the application of `caseList`/`caseList'`. Since our inliner currently
+does not inline recursive bindings, the additional binding will end up not being
+inlined, which can lead to significant performance penalty.
+-}
+
+-- | Matching on the given `List`.
+caseList
+  :: forall a r
+   . (UnsafeFromData a)
+  => (() -> r)
+  -- ^ Nil case
+  -> (a -> List a -> r)
+  -- ^ Cons case
+  -> List a
+  -> r
+-- See Note [Making the cons case non-strict in caseList and caseList']
+caseList n ~c (List l) = B.caseList @BuiltinData n (\x -> c (unsafeFromBuiltinData x) . List) l
```

Comment:
Use `coerce` over the `B.caseList` application to avoid binding `l` to save on an apply-lambda pair?

Reply:
For some reason, doing what you suggested make the mutual recursion appear again. I'll make a branch to demonstrate.

This is quite annoying - we should find a way to inline recursive bindings.